src/backend/optimizer/path/allpaths.c

```diff
@@ -2125,16 +2125,25 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			if (subplan->flow->locustype != CdbLocusType_SegmentGeneral)
+				cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+
```

Comment:
can we save the built plan to not planning it again, as is done in the case of Shared Scan?

Reply:
Do you propose to move the condition from line 2128 to line 2142?

Reply:
I mean is that we can save the plan in `cteplaninfo->shared_plan` without processing in `prepare_plan_for_sharing` and then use it without processing in `share_prepared_plan`.

Reply:
something like
```c
...
cteplaninfo->shared_plan = subplan;
...
if (cteplaninfo->shared_plan->flow->locustype != CdbLocusType_SegmentGeneral)
    subplan = share_prepared_plan(cteroot, prepare_plan_for_sharing(cteroot, cteplaninfo->shared_plan));
...
```
?

Reply:
maybe. But I was thinking about something like this:
```c
subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive, tuple_fraction, &subroot, config);

if(subplan->flow->locustype == CdbLocusType_SegmentGeneral){
   cteplaninfo->shared_plan = subplan;
}else{
   cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
}
```
and
```c
/*
 * Create another ShareInputScan to reference the already-created
 * subplan.
 */
 if(cteplaninfo->shared_plan->flow->locustype == CdbLocusType_SegmentGeneral){
    subplan = cteplaninfo->shared_plan;
 }else{
    subplan = share_prepared_plan(cteroot, cteplaninfo->shared_plan);
 }
 subroot = cteplaninfo->subroot;
```

Reply:
> ```c
>  if(cteplaninfo->shared_plan->flow->locustype == CdbLocusType_SegmentGeneral){
> ```

Is `flow->locustype` same for 

> ```c
>    cteplaninfo->shared_plan = subplan;
> ```

and


> ```c
>    cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
> ```

?

Reply:
what do you mean?

Reply:
Are you sure `prepare_plan_for_sharing` will save `flow->locustype` to `cteplaninfo->shared_plan`?

Reply:
yes https://github.com/arenadata/gpdb/blob/17c7ac54d14f2bc3d2878165d563b84b175e460e/src/backend/optimizer/plan/planshare.c#L224

Reply:
Okay, but I still suggest checking the `locustype` on `SegmentGeneral` only once, not twice.

Reply:
> What if we rename the variable to `plan` as it's not used outside the changed `set_cte_pathlist()`?

I suggest renaming it to `subplan`, and also correcting the comment in the `CtePlanInfo` structure.

Reply:
Turns out we can't simply reuse `subplan`, because it results in construction of DAG instead of a tree, which is later resolved using Shared Scans by design. Using `copyObject` on the subplan seems to work.